### DISCUSSION TITLE
An example on how to deserialize to a struct or hashmap?

### DISCUSSION THREAD

**Discussion by ultrabug** [gklijs/schema_registry_converter#62] [Q&A] [answered]
Hello,

Thanks for this useful crate :+1: 

I'm a bit new to Rust and I must admit I'm confused on how to actually access some sort of Struct or Hashmap from your lib...

I read Avro messages from Kafka, I'd like a way to access the data inside it (key, value). It looks to me (so far) that I need a way to deserialize a `DecodeResult` into something workable like I would expect from a Struct.

The `avro_rs` crate exposes a nice `Reader` struct to deserialize the values from Avro but it doesn't look to work well with DecodeResult.

The gist of my question is, using this crate, how one is supposed to get a workable data object to play with?

PS: I'm using easy async.

```rust
loop {
        match consumer.recv().await {
            Err(e) => warn!("Kafka error: {}", e),
            Ok(m) => {
                let decoded = decoder.decode(m.payload()).await;
                match decoded {
                    Ok(v) => {
                        // this works tho I dunno how to access the actual data inside
                        println!("name {:?}", v.name.unwrap());
                        // what should I do to transform v.value into a struct deserialized by avro_rs?
                    },
                    Err(e) => println!("err {:?}", e),
                }
            }
        };
    }
```

Thanks for your help, I'll sure propose a documentation update once I nail it down :1st_place_medal:

**Comment by gklijs** [comment, marked as answer]
It depends a lot on whether you know the schema of the data up front. If you don't you get back a `Value::Record` which contains the values, but can be tedious to get the information from. Like in https://github.com/gklijs/schema_registry_converter/blob/95f48fbea5f3c6ed5fd3c937689517313eeb1702/src/async_impl/avro.rs#L593 you then would need to match for the vector to contain some field with the expected name, and the expected kind of value, then read the value.

If you do know the schema you can do something similar to https://github.com/gklijs/schema_registry_converter/blob/95f48fbea5f3c6ed5fd3c937689517313eeb1702/src/async_impl/avro.rs#L613 
using the struct defined https://github.com/gklijs/schema_registry_converter/blob/95f48fbea5f3c6ed5fd3c937689517313eeb1702/test_utils/src/lib.rs#L98
I fully agree this later option is kind of hidden, so I would very much appreciate a PR to make this use more clear.

**Reply by ultrabug** [reply]
Thank you for your reply

> If you do know the schema you can do something similar to ... using the struct defined

I tried to do this since I indeed got a Value::Record object but I always get a `DeserializeValue("incorrect value of type: String")` which puzzles me a bit since I implemented a struct like this:

```rust
#[derive(Debug, Deserialize)]
struct Msg {
    tag_id: i32,
    action_id: i8,
}
```

My understanding from the documentation is that the serde Deserialize is supposed to ignore fields that have not been defined in the struct which is what I'm trying to do since I'm only caring about specific fields of the Value::Record. But I'm getting an error from a String field which I'm not interested in anyway...

Am I missing something?

**Reply by gklijs** [reply]
The errors from avro-rs are not always that usefull. Can you maybe share the schema? Alternatively you could try https://github.com/flavray/avro-rs/issues. Another option would be to create producer app, start with a schema with just `tag_id` and `action_id` get it working, and then make it more complex.

**Reply by ultrabug** [reply]
Can I get the fetched schema from a DecodeResult message or from the EasyAvroDecoder? It looks like it's a private field...

**Reply by ultrabug** [reply]
Scratch that, so for now the best I have looks like :

```rust
                // m is a BorrowedMessage from rdkafka
                let decoded = decoder.decode(m.payload()).await.unwrap().value;
                println!("decoded = {:?}", decoded);

                let record = match decoded {
                    Value::Record(v) => v,
                    _ => vec![],
                };

                for (k, v) in record {
                    match k.as_str() {
                        "tag_id" => println!("tag_id = {:?}", from_value::<i32>(&v)),
                        _ => (),
                    };
                    println!("key = {:?} value = {:?}", k, from_value::<String>(&v));
                }
```

Although getting the Record sounds now fine, this sounds very cumbersome to access the data in it... which is my initial help request :pray: please?

Example Record :

```
Record([("oid", String("asdfsdf")), ("uid", String("7873b372-497b-4953-8832-aaaa")), ("fd", Union(String("7873b372-497b-4953-8832-dddd"))), ("td", Union(Null)), ("cd", Union(String("7873b372-497b-4953-8832-sdsdsdsdsd"))), ("platform", String("room")), ("tag_id", Int(6310)), ("action_id", Int(6))])
```

**Reply by gklijs** [reply]
Indeed via this crate it's not easy to get the schema, but you have the data which is almost as good. I would expect if you create a struct with those 5,6? fields it would work. Else indeed you need some matching like you already did. That's kind of the price that's paid for type safety. Something I did before was having an function to extract the fields I wanted and put them in a struct, to prevent deep nested matches. It's niet that pretty, but readable and safe.

**Reply by ultrabug** [reply]
I see thank you, I'll continue a bit on this path and if I come up with something acceptable I'll allow myself a PR to demonstrate it.

**Reply by ultrabug** [reply]
Ok I'd like to add a not to this discussion because as my Rust skills progressed I finally understood what was going on and the problem I was facing come from the avro-rs lib which is actually missing some Union deserializing matches!

I will put up a PR there, this has been a rough ride I must say, and got me dropping the usage of this crate down in the rabbit hole :)

I'll open an issue on other findings related to this crate related to performance

**Reply by ultrabug** [reply]
ofc, if avro-rs accept my proposal, I'll add documentation here to demonstrate how it's done!

**Comment by ultrabug** [comment]
As a follow-up in case someone looks for the answer before a new release of apache avro, here is a sample code that demonstrates how to map Avro data to a struct.

```rust
#[derive(Default, Debug, Deserialize)]
struct Navigation {
  string_field: String,
  nullable_string_field: Option<String>,
}

// messages is rdkafka::message::OwnedMessage (or BorrowedMessage as you wish)
for msg in &messages {
        let decode_result = avro_decoder.decode(msg.payload()).await;
        match decode_result {
            Ok(decoded) => {
                let _navigation: Navigation = match &decoded.value {
                    Value::Record(_) => from_value::<Navigation>(&decoded.value).unwrap(),
                    _ => unimplemented!("from_value did not return a Value::Record"),
                };
            }
            Err(err) => {
                eprintln!("avro decode error {}", err);
                break;
            }
        }
    }
```

**Reply by e-osuna-g** [reply]
I've tried this but with a Partial Struct (just 1 field of many the Record has) an example in [here](https://github.com/apache/avro/blob/master/lang/rust/avro/src/de.rs#L1118), 

```rs
#[derive(Debug, Deserialize, PartialEq, Eq)]
struct KafkaMessage {
    ClientMessageId: String,
    }
    match decoder.decode(msg.payload()).await {
        Ok(v) => {
            println!("value:{:?}", v.value);
            let item = match from_value::<KafkaMessage>(&v.value) {
                Ok(item) => {
                    println!("was decoded {:?}", item)
                },
                Err(e)=> println!("Error on decoding kafkaMessage: {:?}",e)
            };
            v
        },
        Err(e) => panic!("Error getting value: {}", e),
    };
   ```
i always get this error:
``` 
Error on decoding kafkaMessage: DeserializeValue("incorrect value of type: Enum")
```
is there a way to partially getting the record?

**Reply by gklijs** [reply]
You could use the 'raw' values from the record. It seems like the enum is not correct, maybe there is some additional nesting. Hard to say from the example. Maybe [this](https://github.com/gklijs/obm_confluent_blog/blob/f9d38794cf47ce796918d6802dd8056f4d38f16f/command-handler/src/avro_data.rs#L353) help, it's old, but fetching values from the record, instead of using the `from_value` function.

**Reply by e-osuna-g** [reply]
I asked because it seems that it is possible, to create from structure with missing fields with the from_value fn, but i'll try the to_cmt method. thanks

**Reply by gklijs** [reply]
Yes, seems like from_value should work. Maybe it didn't have the expected enum options? Seems like it's actually better/easier to use the from_value fn.
